File: 2023/Day 9 Mirage Maintenance/part1.cpp
#include<iostream>
#include<algorithm>
#include<unordered_map>
#include<fstream>
#include<sstream>
#include<vector>
#include<stack>
using namespace std;
vector<int> solve(string line) {

    vector<int> history;
    vector<int> difference;
    stack<vector<int>> differences;

    string val;
    bool flag = true;
    istringstream ss(line);
    while (ss >> val) {
        history.push_back(stoi(val));
    }
    differences.push(history);

    while (1)
    {   
        for (int i = 0; i < history.size()-1; i++)
        {
           
            difference.push_back(history[i+1] - history[i]);
             
        }
        int count = 0;
        for (int i = 0; i < difference.size(); i++)
        {
            if (difference[i] == 0) {
                count++;
            }
        }
        if (count == difference.size()) {
            break;
        }

        
        differences.push(difference);
        history = {};
        difference = {};

        history = differences.top();
     
    }

int right = 0, left = 0;
while (!differences.empty()) {
    
        vector<int> topVector = differences.top();

        // Part 1
        right += topVector[topVector.size()-1];
        // Part 2
        left = topVector[0] - left;
       
        differences.pop();
    }

return {right, left};
    
    

    
}
int main() {

    ifstream inputFile("D:/Git Hub Repos/Advent of Code/2023/Day 9 Mirage Maintenance/input.txt");
   
    string line;
    int res = 0;
    vector<int> ans;
    int part1 = 0, part2 = 0;
    while (getline(inputFile, line)) {
      
        ans = solve(line);
        part1 += ans[0];
        part2 += ans[1];

    }
    
    
  cout << part1 << endl;
  cout << part2 << endl;

    return 0;
} 

File: 2024/Day 9/part1.py
# def find_freespace(idx):
#     while disk[idx] != -1:
#         idx += 1
#     return idx

with open("D:/Git Hub Repos/Advent of Code/2024/input.txt", 'r') as file:
        line = file.readline().strip()
        
disk = []
file_id = 0

for i, ch in enumerate(line):
    x = int(ch)
    if i % 2 == 0:
        disk += [file_id] * x
        file_id += 1
    else:
        disk += [-1] * x

freespaces = [i for i, x in enumerate(disk) if x == -1]
for space in freespaces:
    while disk[-1] == -1:
        disk.pop()
    if len(disk) <= space: break
    disk[space] = disk.pop()

checksum = 0
for i , x in enumerate(disk):
    checksum +=  i * x
print(checksum)